src/main/java/com/browserstack/automate/ci/jenkins/BrowserStackReportForBuild.java

```diff
@@ -0,0 +1,194 @@
+    }
+
+    private static List<Session> fetchBrowserStackSessions(@Nonnull BrowserStackClient client, @Nonnull String buildId) {
+        List<Session> browserStackSessions = new ArrayList<Session>();
+        try {
+            browserStackSessions.addAll(client.getSessions(buildId));
+        } catch (BuildNotFound bnfException) {
+            log(logger, "No build found while fetching sessions for the buildId: " + buildId);
+        } catch (BrowserStackException bstackException) {
+            log(logger, "BrowserStackException occurred while fetching sessions: " + bstackException.toString());
+        }
+
+        return browserStackSessions;
+    }
+
+    private static List<JSONObject> generateSessionsCollection(List<Session> browserStackSessions) {
+        List<JSONObject> sessionsCollection = new ArrayList<JSONObject>();
+        for (Session session: browserStackSessions) {
+            JSONObject sessionJSON = new JSONObject();
+
+            if (session.getName() == null || session.getName().isEmpty()) {
+                sessionJSON.put("name", session.getId());
+            } else {
+                sessionJSON.put("name", session.getName());
+            }
+
+            if (session.getDevice() == null || session.getDevice().isEmpty()) {
+                sessionJSON.put("browser", session.getBrowser());
+            } else {
+                sessionJSON.put("browser", session.getDevice());
+            }
+            sessionJSON.put("os", session.getOs());
+            sessionJSON.put("osVersion", session.getOsVersion());
+            sessionJSON.put("status", session.getBrowserStackStatus());
+
+            if (session.getBrowserStackStatus().equals(session.getStatus())) {
+                sessionJSON.put("userMarked", "UNMARKED");
+            } else {
+                sessionJSON.put("userMarked", session.getStatus());
+            }
+
+
+            // Condition which shouldn't occur if the build is not being reused elsewhere.
+            // But if it happens, the following condition will handle the scenario where
+            // duration is null or empty (running session)
+            if (session.getStatus().equals("running")) {
+                sessionJSON.put("duration", "-");
+            } else {
+                sessionJSON.put("duration", session.getDuration());
+            }
+
+            sessionJSON.put("createdAt", session.getCreatedAt());
+            sessionJSON.put("url", session.getPublicUrl() + "&source=jenkins");
+            sessionsCollection.add(sessionJSON);
+        }
+
+        return sessionsCollection;
+    }
+
+    private static class SessionsSortingComparator implements Comparator<JSONObject> {
+
+        @Override
+        public int compare(JSONObject sessionOne, JSONObject sessionTwo) {
+            // possible values for user_marked: failed, passed and UNMARKED, thus changing all to lowercase
+            String sessionOneUserMarked = sessionOne.getString("userMarked").toLowerCase();
```

Comment:
can it be null?

Reply:
nope. since it fetches the values populated by the prior function.